support-lib/jni/djinni_support.cpp

```diff
@@ -239,8 +239,13 @@ jint JniEnum::ordinal(JNIEnv * env, jobject obj) const {
 
 LocalRef<jobject> JniEnum::create(JNIEnv * env, jint value) const {
     LocalRef<jobject> values(env, env->CallStaticObjectMethod(m_clazz.get(), m_staticmethValues));
+    jniExceptionCheck(env);
     DJINNI_ASSERT(values, env);
-    return LocalRef<jobject>(env, env->GetObjectArrayElement(static_cast<jobjectArray>(values.get()), value));
+    LocalRef<jobject> result(env,
+                             env->GetObjectArrayElement(static_cast<jobjectArray>(values.get()),
+                                                        value));
+    jniExceptionCheck(env);
+    return std::move(result);
```

Comment:
Ah, thanks for taking the time to research that.  Good to know, and it does make sense that the code has to be legal without RVO, and can't move without RVO.  The standard needs to allow for things like move/copy ctors with unusual semantics or side-effects, and thus can't do substitutions or omissions except where explicitly allowed to.

That makes the best practice a little easier to follow, though still not easy.  For a well-behaved type which is copyable, making it const won't kill RVO (though it's a bit tricky to be sure where RVO will happen).  And for a move-only type, the compiler failure will remind you to remove the const (though an unwary programmer might add a move at the same time like I did).

Good discussion.  I learned a lot.